Vulkan: Implement pipeline UID cache

stenzek opened the pull request:
This stores enough information to recreate the pipeline, including the shader UIDs, blend/depth/rasterization state, primitive and vertex format.

Should get rid of shader compiling stutter on **second boots of games** to the same degree that it is mitigated on the D3D/GL backends, if not even better, because we won't be hitting shader recompiles due to state changes (all these are tracked).

Especially makes a difference for AMD users, where the Vulkan pipeline cache does not appear to be populated with any data, so every boot resulted in full recompiles (apart from the initial GLSL generation being skipped).

Also adds validation of pipeline objects, I assume drivers should be doing this already, but just in case.

**NOTE:** Currently if a user flips between MSAA/SSAA on/off, this will increase the boot time of games because all variations of shaders will be generated. This is an issue that affects all backends, when/if we decouple MSAA state from the UIDs it will sort itself out.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/4449)
<!-- Reviewable:end -->

lioncash commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/VideoBackends/Vulkan/ObjectCache.cpp, line 162 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTMSxBGhqvET1CnhTE:-KWTMSxCrITd6YvnLdWa:bd1xqgt) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/ObjectCache.cpp#L162)):*
> ```cpp
> }
> 
> VkPipeline ObjectCache::GetPipeline(const PipelineInfo& info, bool* was_cache_miss)
> ```

This should probably return a tuple or pair containing vkPipeline and a boolean instead of using an out parameter.

---

*[Source/Core/VideoBackends/Vulkan/ObjectCache.cpp, line 385 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTMHbDu4gUr9-MGVUs:-KWTMHbDu4gUr9-MGVUt:btsosvy) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/ObjectCache.cpp#L385)):*
> ```cpp
>   u32 device_id;
>   u8 uuid[VK_UUID_SIZE];
> };
> ```

@degasus Did you update the buildbot to GCC 5 yet? This is a prime candidate for having 

```cpp
static_assert(std::is_trivially_copyable<VK_PIPELINE_CACHE_HEADER>(), "Header must be trivially copyable");
```
after it.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 123 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTM96Ke4SPJaBBkHsX:-KWTM96LmeLEBmUHWWf9:b-tkvqov) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L123)):*
> ```cpp
> void StateTracker::LoadPipelineUIDCache()
> {
>   class PipelineInserter : public LinearDiskCacheReader<SerializedPipelineUID, u32>
> ```

This should be `final` if it's not going to be extended.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 126 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTMC9SLleXiZa9xeNJ:-KWTMC9SLleXiZa9xeNK:b4ryxev) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L126)):*
> ```cpp
>   {
>   public:
>     PipelineInserter(StateTracker* this_ptr_) : this_ptr(this_ptr_) {}
> ```

`explicit PipelineInserter`

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 167 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTN-6hTx79dPUvZ-6I:-KWTN-6j_0cTqM8N0d-5:blba8vo) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L167)):*
> ```cpp
>   // Do we have a vertex format created that matches the declaration?
>   auto vertex_format_map = VertexLoaderManager::GetNativeVertexFormatMap();
>   auto iter = vertex_format_map->find(uid.vertex_decl);
> ```

These and the following if statement should probably be their own function.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 174 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTPWCQWpWVa3BrLZh9:-KWTPWCQWpWVa3BrLZhA:b-vmb12q) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L174)):*
> ```cpp
>     // Requires the new operator because we need to upcast it.
>     auto ipair = vertex_format_map->emplace(
>         uid.vertex_decl, std::unique_ptr<NativeVertexFormat>(new VertexFormat(uid.vertex_decl)));
> ```

You can use `std::make_unique` here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449)*
<!-- Sent from Reviewable.io -->

stenzek commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[Source/Core/VideoBackends/Vulkan/ObjectCache.cpp, line 162 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTMSxBGhqvET1CnhTE:-KWrbXUyXAx0QRJWNR3D:bsxzlof) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/ObjectCache.cpp#L162)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This should probably return a tuple or pair containing vkPipeline and a boolean instead of using an out parameter.
</details>
Done, although I added a second method for this case, as all but one of the callers don't care about whether it was a cache hit or miss.

---

*[Source/Core/VideoBackends/Vulkan/ObjectCache.cpp, line 385 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTMHbDu4gUr9-MGVUs:-KWrbtfRyEHCC5bwgAau:bl0gg8g) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/ObjectCache.cpp#L385)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> @degasus Did you update the buildbot to GCC 5 yet? This is a prime candidate for having 
> 
> ```cpp
> static_assert(std::is_trivially_copyable<VK_PIPELINE_CACHE_HEADER>(), "Header must be trivially copyable");
> ```
> after it.
</details>
Done. Looks like the builder isn't updated though, so I'll have to add a guard around it for now.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 123 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTM96Ke4SPJaBBkHsX:-KWrbx3OSgkPQjAKGiQt:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L123)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This should be `final` if it's not going to be extended.
</details>
Done.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 126 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTMC9SLleXiZa9xeNJ:-KWrbxlkw4LckEZ41h9K:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L126)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `explicit PipelineInserter`
</details>
Done.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 167 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTN-6hTx79dPUvZ-6I:-KWrbzEqsRLzNYM3ePwW:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L167)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> These and the following if statement should probably be their own function.
</details>
Done.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 174 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KWTPWCQWpWVa3BrLZh9:-KWrbzj7zAoK0UJqA1Z3:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ebb3fe618f97f2a6fdac6b1b7b4d4eef1d40bc6f/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L174)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You can use `std::make_unique` here.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449)*
<!-- Sent from Reviewable.io -->

degasus commented:
Sounds fine.

---

Reviewed 2 of 5 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[Source/Core/VideoBackends/Vulkan/ObjectCache.cpp, line 291 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KXeGUBebtSVPlPFJOGD:-KXeGUBebtSVPlPFJOGE:b-7muqvf) ([raw file](https://github.com/dolphin-emu/dolphin/blob/1c51592e5119a008d2bb01185197da4d563abc26/Source/Core/VideoBackends/Vulkan/ObjectCache.cpp#L291)):*
> ```cpp
> VkPipeline ObjectCache::GetPipeline(const PipelineInfo& info)
> {
>   auto iter = m_pipeline_objects.find(info);
> ```

return GetPipelineWithCacheResult(info).first;

Do you need both functions?

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 140 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KXeHmR-ZV9YpDTEmznu:-KXeHmR-ZV9YpDTEmznv:b-fvgniy) ([raw file](https://github.com/dolphin-emu/dolphin/blob/1c51592e5119a008d2bb01185197da4d563abc26/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L140)):*
> ```cpp
>   // OpenAndRead calls Close() first, which will flush all data to disk when reloading.
>   // This assertion must hold true, otherwise data corruption will result.
>   m_uid_cache.OpenAndRead(filename, inserter);
> ```

Feel free to change the API of  our LinearDiskCache to use a std::function ;)

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 154 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KXeIoJQGhMUTZTydxaD:-KXeIoJQGhMUTZTydxaE:b-jn4nqv) ([raw file](https://github.com/dolphin-emu/dolphin/blob/1c51592e5119a008d2bb01185197da4d563abc26/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L154)):*
> ```cpp
>   sinfo.ps_uid = m_ps_uid;
>   sinfo.primitive_topology = info.primitive_topology;
>   sinfo.bbox_enabled = m_bbox_enabled;
> ```

Out of curiosity, is this the global config option, for the hardware support, or the global state? I wonder why this isn't within the ps_uid.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 923 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KXeJfIFMZ15TljUaG5d:-KXeJfIFMZ15TljUaG5e:b-31x0r2) ([raw file](https://github.com/dolphin-emu/dolphin/blob/1c51592e5119a008d2bb01185197da4d563abc26/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L923)):*
> ```cpp
>     // the existing state, since we don't want to break the next draw.
>     PipelineInfo temp_info = GetAlphaPassPipeline(m_pipeline_state);
>     auto result = g_object_cache->GetPipelineWithCacheResult(temp_info);
> ```

Should there be a function for this code as it's used twice?

    auto result = g_object_cache->GetPipelineWithCacheResult(STATE);
    m_pipeline_object = result.first;
    if (m_pipeline_object == VK_NULL_HANDLE)
      return false;
 
    // Add to the UID cache if it is a new pipeline.
    if (!result.second)
      AppendToPipelineUIDCache(STATE);

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-:-KXeKU1zRuzQXtlDnqsP:brx1mkd)*
<!-- Sent from Reviewable.io -->

stenzek commented:
Review status: 3 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[Source/Core/VideoBackends/Vulkan/ObjectCache.cpp, line 291 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KXeGUBebtSVPlPFJOGD:-KXed5wd2BN8gdpDbptz:bhta0ly) ([raw file](https://github.com/dolphin-emu/dolphin/blob/1c51592e5119a008d2bb01185197da4d563abc26/Source/Core/VideoBackends/Vulkan/ObjectCache.cpp#L291)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary><blockquote>

return GetPipelineWithCacheResult(info).first;

Do you need both functions?
</blockquote></details>

I've simplified the GetPipeline() method. I originally just had it as an out parameter of GetPipeline, but following Lioncash's recommendation I used a pair return instead. This was the cleanest way I could think of accomplishing this without modifying all the callers (which all except one don't care about whether it was a cache miss).

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 140 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KXeHmR-ZV9YpDTEmznu:-KXedNKX9Lou_T4Rap_X:beqaedo) ([raw file](https://github.com/dolphin-emu/dolphin/blob/1c51592e5119a008d2bb01185197da4d563abc26/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L140)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary><blockquote>

Feel free to change the API of  our LinearDiskCache to use a std::function ;)
</blockquote></details>

Sounds like a good idea, outside of the scope of this PR I think.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 154 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KXeIoJQGhMUTZTydxaD:-KXedYabkPFYeynoAS9H:bvwocbf) ([raw file](https://github.com/dolphin-emu/dolphin/blob/1c51592e5119a008d2bb01185197da4d563abc26/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L154)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary><blockquote>

Out of curiosity, is this the global config option, for the hardware support, or the global state? I wonder why this isn't within the ps_uid.
</blockquote></details>

It's an variable here because I use a different pipeline layout depending on whether bbox is enabled or not (avoiding the need to bind the shader storage buffer). But you're right about it being part of the ps_uid, I've simplified things and used that field instead. Could probably remove the boolean from StateTracker too, later on.

---

*[Source/Core/VideoBackends/Vulkan/StateTracker.cpp, line 923 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-KXeJfIFMZ15TljUaG5d:-KXefVgXgJxjZ07yoDS5:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/1c51592e5119a008d2bb01185197da4d563abc26/Source/Core/VideoBackends/Vulkan/StateTracker.cpp#L923)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary><blockquote>

Should there be a function for this code as it's used twice?

    auto result = g_object_cache->GetPipelineWithCacheResult(STATE);
    m_pipeline_object = result.first;
    if (m_pipeline_object == VK_NULL_HANDLE)
      return false;
 
    // Add to the UID cache if it is a new pipeline.
    if (!result.second)
      AppendToPipelineUIDCache(STATE);
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449)*
<!-- Sent from Reviewable.io -->

degasus commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4449#-:-KXehl_OTmSTQE7B7wGV:bnfp4nl)*
<!-- Sent from Reviewable.io -->
